src/libPMacc/include/random/Random.hpp

```diff
@@ -0,0 +1,106 @@
+/**
+ * Copyright 2015 Alexander Grund
+ *
+ * This file is part of libPMacc.
+ *
+ * libPMacc is free software: you can redistribute it and/or modify
+ * it under the terms of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * libPMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with libPMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "types.h"
+#include "random/RNGState.hpp"
+#include <boost/utility/result_of.hpp>
+
+namespace PMacc {
+namespace random {
+
+    /**
+     * Random Number Generator. Functor that returns a random number per call
+     *
+     * Default implementation assumes a RNGHandle
+     */
+    template<
+        class T_Distribution,
+        class T_RNGMethod,
+        class T_RNGStateOrHandle = typename T_RNGMethod::StateType*
```

Comment:
> Is typdef typename T_RNGMethod::StateType*::RNGState Foo a allowed operation

There is a specialisation for pointer types so this will never be the case